cluster-autoscaler/utils.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"time"
+
+	kube_api "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// UnscheduledPodLister list unscheduled pods
+type UnscheduledPodLister struct {
+	podLister *cache.StoreToPodLister
+}
+
+// List returns all unscheduled pods.
+func (unscheduledPodLister *UnscheduledPodLister) List() ([]*kube_api.Pod, error) {
+	//TODO: Extra filter based on pod condition.
+	return unscheduledPodLister.podLister.List(labels.Everything())
+}
+
+// NewUnscheduledPodLister returns a lister providing pods that failed to be scheduled.
+func NewUnscheduledPodLister(kubeClient *kube_client.Client) *UnscheduledPodLister {
+	// watch unscheduled pods
+	selector := fields.ParseSelectorOrDie("spec.nodeName==" + "" + ",status.phase!=" +
+		string(kube_api.PodSucceeded) + ",status.phase!=" + string(kube_api.PodFailed))
+	podListWatch := cache.NewListWatchFromClient(kubeClient, "pods", kube_api.NamespaceAll, selector)
+	podLister := &cache.StoreToPodLister{Store: cache.NewStore(cache.MetaNamespaceKeyFunc)}
+	podReflector := cache.NewReflector(podListWatch, &kube_api.Pod{}, podLister.Store, time.Hour)
+	podReflector.Run()
+
+	return &UnscheduledPodLister{
+		podLister: podLister,
+	}
+}
+
+// NewNodeLister builds a node lister.
+func NewNodeLister(kubeClient *kube_client.Client) *cache.StoreToNodeLister {
+	listWatcher := cache.NewListWatchFromClient(kubeClient, "nodes", kube_api.NamespaceAll, fields.Everything())
```

Comment:
I think should take into account only nodes in state `Ready`. Unhealthy nodes are not interesting to us.